docs/understand/programming_model.rst

```diff
@@ -111,46 +160,213 @@ typically look the following:
-usually isn't exploited from the width of the built-in vector types, but via the
-thread id constants ``threadIdx.x``, ``blockIdx.x``, etc. For more details,
-refer to :ref:`inherent_thread_model`.
+In HIP, lanes of the SIMD architecture are fed by mapping threads of a SIMT
+execution, one thread down each lane of an SIMD engine. Execution parallelism
+usually isn't exploited from the width of the built-in vector types, but across multiple threads via the thread ID constants ``threadIdx.x``, ``blockIdx.x``, etc. 
 
-Heterogeneous programming
-=========================
+.. _inherent_thread_model:
 
-The HIP programming model assumes two execution contexts. One is referred to as
-*host* while compute kernels execute on a *device*. These contexts have
-different capabilities, therefor slightly different rules apply. The *host*
-execution is defined by the C++ abstract machine, while *device* execution
-follows the HIP model, primarily defined by SIMT. These execution contexts in
-code are signified by the ``__host__`` and ``__device__`` decorators. There are
-a few key differences between the two:
+Inherent thread model
+=====================
 
-* The C++ abstract machine assumes a unified memory address space, meaning that
-  one can always access any given address in memory (assuming the absence of
-  data races). HIP however introduces several memory namespaces, an address
-  from one means nothing in another. Moreover, not all address spaces are
-  accessible from all contexts.
+The SIMT nature of HIP is captured by the ability to execute user-provided
+device programs, expressed as single-source C/C++ functions or sources compiled
+online/offline to binaries, in bulk.
 
-  If one were to look at :ref:`cdna2_gcd` and inside the :ref:`cdna3_cu`,
-  every Compute Unit has an instance of storage backing the namespace
-  ``__shared__``. Even if the host were to have access to these regions of
-  memory, the performance benefits of the segmented memory subsystem are
-  supported by the inability of asynchronous access from the host.
+All :ref:`threads <programming_model_simt>` of a kernel uniquely identified by a set of integral values, or thread IDs.
+The set of integers identifying a thread relate to the hierarchy in which threads execute.
 
-* Not all C++ language features map cleanly to typical device architectures,
-  some are very expensive (meaning slow) to implement on GPU devices, therefor
-  they are forbidden in device contexts to avoid users tapping into features
-  that unexpectedly decimate their program's performance. Offload devices targeted
-  by HIP aren't general purpose devices, at least not in the sense a CPU is.
-  HIP focuses on data parallel computations and as such caters to throughput
-  optimized architectures, such as GPUs or accelerators derived from GPU
-  architectures.
+The thread hierarchy inherent to how AMD GPUs operate is depicted in
+:numref:`inherent_thread_hierarchy`.
```

Comment:
This reference does not render correctly.

Reply:
Fixed.